### PR TITLE
Cosmos DB - availability strategy fix with query

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

This PR adds option to do hedging via availability strategy not just for point operations but also query operations.

When adding test coverage for this feature a few issues with regards to diagnostics in queries have been identified and also fixed in this PR:
- Cardinality of CosmosDiagnosticsContext / Trace spans etc.
 - Before this PR a single query operation in the SDK returning a `CosmosPagedFlux` was always captured in a single CosmosDiagnosticsContext instance / and a single tracing span. The guiding principle is that there should be one ComsosDiagnosticsContext and trace span per user operation in the SDK. For queries it is not super clear whether you should treat creating a query and full draining it as a single user interaction or whether processing each page returned from the `CosmosPagedFlux` should be treated as the user interaction. For single partition queries returning  only few records it is the same - but especially for long-running queries or for queries across dozens/hundreds of partitions it makes a big difference. When having only one CosmosDiagnosticsCOntext per query it means metrics/logging etc. will only happen at the end-of the query's lifecylce - it seems preferrable to instead emit them per page to allow monitoring while the long running query is still processing as well. After some offline discussions we decided to switch to emitting CosmosDiagnosticsContext per page - and this is also included in this PR.
- Empty Page handling: This PR also fixes a gap in diagnostics for queries returning empty pages (either because a cross partition query does not return results for some of the partitions or when a query does not return any results). The per request statistics for these empty pages was missing - and as a result they would not have been reflected in metrics/diagnostics etc. This gap is also addressed in this PR
- SequenceNumber - this PR adds a new property `CosmosDiagnosticsContext.getSequnceNumber` which is only populated for feed operations - the combination of `CorrelatedActivityId` and `SequenceNumber` will be able to reconstruct the order in which operations/traces etc. have happened.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 1]
```diff
@@ -661,7 +661,8 @@ <T> CosmosPagedFlux<T> readAllItems(CosmosQueryRequestOptions options, Class<T>
                 client,
                 queryOptionsAccessor.getQueryNameOrDefault(requestOptions, this.readAllItemsSpanName),
                 requestOptions.getConsistencyLevel(),
-                client.getEffectiveDiagnosticsThresholds(queryOptionsAccessor.getDiagnosticsThresholds(requestOptions)));
+                client.getEffectiveDiagnosticsThresholds(queryOptionsAccessor.getDiagnosticsThresholds(requestOptions)),
+                queryOptionsAccessor.getDiagnosticsFactoryResetCallbackReference(requestOptions));
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:665]: do we need to also set the context supplier here: 
```
           queryOptionsAccessor.setDiagnosticsContextSupplier(
                options,
                () -> pagedFluxOptions.getDiagnosticsContextSnapshot());
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 2]
```diff
@@ -5407,5 +5640,10 @@ public void merge(RequestOptions requestOptions) {
                 }
             }
         }
+
+        public void reset() {
+            this.createdDiagnostics.clear();
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:5645]: Trying to understand the design here:

- For every response (mapped to a request) a merge of the `CosmosDiagnostics` instance is done (could be cancellation or an error response as well).
- A `reset` is done when a given "operation". For query - 1 "operation" is when 1 `FeedResponse` is obtained. Preceding a `reset` is when you create a new `CosmosDiagnosticContext` instance with a new sequence no.

**Review by xinlian12 — APPROVED** [review]
LGTM, great changes, thanks 👍

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 3]
```diff
@@ -951,15 +957,30 @@ private <T> Flux<FeedResponse<T>> createQuery(
...
+        final ScopedDiagnosticsFactory diagnosticsFactory = new ScopedDiagnosticsFactory(this);
+        qryOptAccessor.setDiagnosticsFactoryResetCallback(nonNullQueryOptions, () -> diagnosticsFactory.reset());
+        return
+            ObservableHelper.fluxInlineIfPossibleAsObs(
+                                () -> createQueryInternal(
+                                    diagnosticsFactory, resourceLink, sqlQuery, nonNullQueryOptions, klass, resourceTypeEnum, queryClient, correlationActivityId, isQueryCancelledOnTimeout),
+                                invalidPartitionExceptionRetryPolicy
+                            )
+                            .flatMap(result -> {
+                                diagnosticsFactory.merge(nonNullRequestOptions);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:972]: just for personal understanding -
I guess by theory, we do not really need the merge here any more? because now we reset for each page?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 4]
```diff
@@ -1089,7 +1098,8 @@ <T> Function<CosmosPagedFluxOptions, Flux<FeedResponse<T>>> queryChangeFeedInter
                 spanName,
                 null,
                 client.getEffectiveDiagnosticsThresholds(
-                    cfOptionsAccessor.getDiagnosticsThresholds(cosmosChangeFeedRequestOptions)));
+                    cfOptionsAccessor.getDiagnosticsThresholds(cosmosChangeFeedRequestOptions)),
+                null);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1102]: why for changeFeed, we do not follow the per trace per page logic?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:972]: You are amazing - yes and no - you are right that this is the wrong place for merge - causing the majority of the remaining test failures.

Reset is just Creating a new CosmosDiagnosticsContext - not merging in the CosmosDiagnostics instances - so, merge is still needed - but it should be wired up the same way as reset. Working on that right now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/IDocumentQueryClient.java [hunk 5]
```diff
@@ -2,14 +2,26 @@
 // Licensed under the MIT License.
 package com.azure.cosmos.implementation.query;
 
+import com.azure.cosmos.CosmosEndToEndOperationLatencyPolicyConfig;
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/IDocumentQueryClient.java:5]: Fixed

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/FaultInjectionWithAvailabilityStrategyTests.java [hunk 6]
```diff
@@ -1531,48 +1528,262 @@ public void writeAfterCreation(
...
+                + "')";
+
+        Function<ItemOperationInvocationParameters, String> crossPartitionWithAggregatesAndOrderByQueryGenerator = (params) ->
+            "SELECT DISTINCT c.id FROM c WHERE CONTAINS (c.id, '"
+                + params.idAndPkValuePair.getLeft()
+                + "')";
+
+        Function<ItemOperationInvocationParameters, String> crossPartitionEmptyResultQueryGenerator = (params) ->
+            "SELECT * FROM c WHERE CONTAINS (c.id, 'NotExistingId')";
+
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/FaultInjectionWithAvailabilityStrategyTests.java:1683]: Agreed for readMany and readAll - will add some coverage for those

All kind of query flavors seems too much here - I have intentionally covered distinct with order by because that uses a different code path in query pipeline - believe that is a reasonable compromise

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: I think addressed with the FeedOperationState refactoring

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:5677]: Correct - The DiagnosticsFactory will capture all CosmosDiagnostics instances - when reset is called it clears the list of captured "CosmosDianostics instances created" - merge will make sure all currently captured diagnostics instances will be mapped to the current CosmosDiagnosticsContext

The overall goal for PagedFlux is
Ever FeedResponse bubbled up to the application will have its own ComosDiagnosticsContext instance - which will have potentially multipel CosmosDiagnostics instances (for example 1 for query plan, 1 for request to first region, one for hedign request to second region). With aggregates One FeedResponse bubbled up to application could also mean we had to reach out to multiple partitions internally (but those would be different request statistics within one ComosDiagnostics)

To achieve this we have the diagnostics factory capturing all diagnostics being created - whenever we bubble up a new feedresponse we merge (map the current diagnostics context to all previously captured diagnostics) and then clear the list of captured contexts)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: I have eft Changefeed untouched for now - let's discuss in person

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/FaultInjectionWithAvailabilityStrategyTests.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/FaultInjectionWithAvailabilityStrategyTests.java:1683]: Will add the tests for readAll and readMany in separate PR.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/SparkBridgeInternal.scala [hunk 7]
```diff
@@ -56,10 +61,12 @@ private[cosmos] object SparkBridgeInternal {
   ): List[PartitionKeyRange] = {
     val pkRanges = new ArrayBuffer[PartitionKeyRange]()
 
+
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/SparkBridgeInternal.scala:65]: nit:extra line

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ClientMetricsTest.java [hunk 8]
```diff
@@ -249,7 +262,7 @@ public void createItem() throws Exception {
                     this.validateMetrics(
                         expectedOperationTag,
                         expectedRequestTag,
-                        1,
+                        0,
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ClientMetricsTest.java:265]: why the change here

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/TracerUnderTest.java [hunk 9]
```diff
@@ -102,21 +111,29 @@ public synchronized void end(String statusMessage, Throwable error, Context cont
         this.currentSpan.setError(error);
         this.currentSpan.setStatusMessage(statusMessage);
 
+        String spanName = this.currentSpan.name;
+
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/TracerUnderTest.java:116]: nit: extra line

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 10]
```diff
@@ -1548,22 +1548,27 @@ public <T> CosmosPagedFlux<T> readAllItems(
                 : options;
         requestOptions.setPartitionKey(partitionKey);
 
+
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1552]: nit: extra line

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedOperationState.java [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            .CosmosDiagnosticsContextHelper
+            .getCosmosDiagnosticsContextAccessor();
+
+    private final CosmosAsyncClient cosmosAsyncClient;
+    private final CosmosDiagnosticsThresholds thresholds;
+
+
+    private final AtomicReference<CosmosDiagnosticsContext> ctxHolder;
+
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedOperationState.java:38]: nit: extra line

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ClientMetricsTest.java (hunk 8 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ClientMetricsTest.java:265]: Random test flakiness - when backend indeed returns sometimes 0 RU.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedOperationState.java (hunk 11 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/FeedOperationState.java:38]: Fixed

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 10 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1552]: Fixed

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/SparkBridgeInternal.scala (hunk 7 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/SparkBridgeInternal.scala:65]: Fixed
